Implementing Update method to ServicePlan resource

RDavisHPE opened the pull request:
Implementing Update method for ServicePlan resource to handle updates to existing service plans in Morpheus as part of the Terraform provider.
- Implemented Update method in internal/subproviders/morpheus/resources/serviceplan/resource.go
- Implemented acceptance tests in update_test.go

Also updated `ValidateConfig` to allow unknown variables for custom_max_memory, custom_max_storage, add_volumes and custom_cores as this allows for use of variables in the acceptance tests without triggering validation.

RDavisHPE committed:
Implementing Update method to ServicePlan resource

Implementing Update method for ServicePlan resource to handle updates to
existing service plans in Morpheus as part of the Terraform provider.
- Implemented Update method in internal/subproviders/morpheus/resources/serviceplan/resource.go
- Implemented acceptance tests in update_test.go

matthew-c-hpe commented on internal/subproviders/morpheus/resources/serviceplan/resource.go:
```diff
@@ -398,16 +503,163 @@ func (r *Resource) Create(
+		return
+	}
+
+	if state.Id.IsNull() || state.Id.IsUnknown() {
```
Assuming Import and Create have been coded correctly, it should be impossible to get to the point where ID isn't set when we perform an update, so this check is redundant.

Create and Import for Service Plan have enough guarantees to make sure ID isn't null or unknown.

RDavisHPE replied:
Good catch - removed this check 👍

matthew-c-hpe commented on internal/subproviders/morpheus/resources/serviceplan/resource.go:
```diff
@@ -398,16 +503,163 @@ func (r *Resource) Create(
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		var pricesets []sdk.AddServicePlansRequestServicePlanPriceSetsInner
```
You should be able to skip this step and convert straight to `UpdateServicePlansRequestServicePlanPriceSetsInner`

RDavisHPE replied:
Surprised I missed that one! Also noticed that the struct takes `AddServicePlanRequestServicePlanPriceSetsInner`: 
```go
type UpdateServicePlansRequestServicePlan struct { /* ... */
	// List of price sets to include in service plan
	PriceSets            []AddServicePlansRequestServicePlanPriceSetsInner `json:"priceSets,omitempty"`
 }
```

matthew-c-hpe replied:
Yeah - that's down to how OpenAPI Generator will reuse generated nested structs with equivalent shapes. Can be pretty annoying and confusing and potentially prone to stuff breaking (names of struct types).

matthew-c-hpe commented on internal/subproviders/morpheus/resources/serviceplan/update_test.go:
```diff
@@ -0,0 +1,446 @@
+}
+
+// Helper function to merge config variables
+func mergeConfigVars(base config.Variables, overrides config.Variables) config.Variables {
```
You don't need to merge the config vars; the point of using default values in the variables you're using in your config is so that you can simply provide a single `config.Variable` for each step and it'll detect changes only on that. For the rest, it'll fall back to your default values. So you can remove this and reduce the base Config vars to something like stuart does, i.e. just a single variable.

RDavisHPE replied:
Good catch - removed that function and set the only base config var to `name` 👍

RDavisHPE committed:
Simplifying Service Plan update and tests
- Removed unnecessary id check in Update function
- Removed unnecessary iteration through Price Sets in Update function
- Removed unnecessary base config variables in update_test.go and removed
  mergeConfigVars.

RDavisHPE committed:
Removing replace SDK statement for local testing

matthew-c-hpe commented on internal/subproviders/morpheus/resources/serviceplan/update_test.go:
```diff
@@ -0,0 +1,440 @@
+					resource.TestCheckResourceAttr("hpe_morpheus_service_plan.test", "config_ranges.max_sockets", "20"),
+					resource.TestCheckResourceAttr("hpe_morpheus_service_plan.test", "config_ranges.min_cores_per_socket", "2"),
+					resource.TestCheckResourceAttr("hpe_morpheus_service_plan.test", "config_ranges.max_cores_per_socket", "20"),
+				),
```
Let's add another step to test an update of `storage_size_type` and `memory_size_type`, or include them in here. We want to make sure there's no side effects from updating them. See how the API behaves first though, separate to the acceptance tests via curl or Morpheus CLI

RDavisHPE replied:
Added to final apply (could also think of adding a check that max_storage and max_memory haven't changed in their respective types individual checks?)

Going off experimentation they don't change with changing the types so this is probably enough :+1:

RDavisHPE replied:
Added to final apply. If we wanted we could also have:
```go
{ // memory_size_type change isolation test - verify no side effects on max_memory
	Config: baseConfigText,
	ConfigVariables: config.Variables{
		"name":             config.StringVariable(uniqueName),
		"memory_size_type": config.StringVariable("gb"),
	},
	Check: resource.ComposeTestCheckFunc(
		// Verify the size type changed
		resource.TestCheckResourceAttr("hpe_morpheus_service_plan.test", "memory_size_type", "gb"),
		// Verify max_memory value unchanged (no side effects)
		resource.TestCheckResourceAttr("hpe_morpheus_service_plan.test", "max_memory", "4294967296"),
	),
	}, { // storage_size_type change isolation test - verify no side effects on max_storage
	Config: baseConfigText,
	ConfigVariables: config.Variables{
		"name":              config.StringVariable(uniqueName),
		"storage_size_type": config.StringVariable("mb"),
	},
	Check: resource.ComposeTestCheckFunc(
		// Verify the size type changed
		resource.TestCheckResourceAttr("hpe_morpheus_service_plan.test", "storage_size_type", "mb"),
		// Verify max_storage value unchanged (no side effects)
		resource.TestCheckResourceAttr("hpe_morpheus_service_plan.test", "max_storage", "0"),
	),
},
```
However through experimentation it seems updating the number is done through the UI so this is unlikely to break. (edit: didn't realize I sent the previous reply)

RDavisHPE committed:
Adding storage_size_type and memory_size_type to final apply update test
and updated documentation to reflect Update has been implemented.

matthew-c-hpe reviewed (approved):
Looks good to me, just update the description to capture the changes to `ValidateConfig`
